File: PaperDeck/Assets/Scripts/Packets/AuthPacket.cs
using PaperDeck.Network;

namespace PaperDeck.Packets
{
    /// <summary>
    /// Received from the server when a player joins the server.
    /// </summary>
    public class AuthPacket : IPacket
    {
        public const long ID = 3782780678832128L;

        /// <summary>
        /// The name of this client's player.
        /// </summary>
        public string PlayerName { get; set; }

        /// <summary>
        /// The ID of this client's player.
        /// </summary>
        public string PlayerID { get; set; }

        /// <inheritdoc cref="IPacket"/>
        public long PacketID => ID;

        /// <inheritdoc cref="IPacket"/>
        public void Handle()
        {
            // Never received from server.
        }
    }

    /// <summary>
    /// The packet IO handler for sending auth packets.
    /// </summary>
    public class AuthPacketIO : IPacketIO<AuthPacket>
    {
        /// <inheritdoc cref="IPacketIO{T}"/>
        public long PacketID => AuthPacket.ID;

        /// <inheritdoc cref="IPacketIO{T}"/>
        public IPacket ReadPacket(DataInput reader)
        {
            // Never sent from server
            throw new System.ApplicationException("Illegal packet received from server!");
        }

        /// <inheritdoc cref="IPacketIO{T}"/>
        public void WritePacket(DataOutput writer, AuthPacket packet)
        {
            writer.Write(packet.PlayerName);
            writer.Write(packet.PlayerID);
        }
    }
}

File: Server/src/main/java/net/whg/paperdeck/players/AuthPacket.java
package net.whg.paperdeck.players;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import net.whg.paperdeck.IOUtils;
import net.whg.we.net.IPacketSender;
import net.whg.we.net.packets.IBinaryPacket;
import net.whg.we.net.packets.IPacketInitializer;
import net.whg.we.net.server.IConnectedClient;

/**
 * When received, authentications a player and adds them to the player list.
 */
public class AuthPacket implements IBinaryPacket
{
    private static final Logger logger = LoggerFactory.getLogger(AuthPacket.class);
    private static final long PACKET_ID = 3782780678832128L;

    /**
     * Initializes received ping packets.
     */
    public static class Initializer implements IPacketInitializer<AuthPacket>
    {
        private final PlayerList playerList;

        public Initializer(PlayerList playerList)
        {
            this.playerList = playerList;
        }

        @Override
        public long getPacketID()
        {
            return PACKET_ID;
        }

        @Override
        public AuthPacket loadPacket(DataInput input, IPacketSender sender) throws IOException
        {
            var playerName = IOUtils.readString(input);
            var playerID = IOUtils.readString(input);
            return new AuthPacket(sender, playerName, playerID, playerList);
        }
    }

    private final IPacketSender sender;
    private final String playerName;
    private final String playerID;
    private final PlayerList playerList;

    /**
     * Creates a new ping packet.
     * 
     * @param sender
     *     - The client who sent this packet.
     * @param playerName
     *     - The name of the player.
     * @param playerID
     *     - The ID of the player.
     * @param playerList
     *     - The list of players to add the client to.
     */
    AuthPacket(IPacketSender sender, String playerName, String playerID, PlayerList playerList)
    {
        this.sender = sender;
        this.playerName = playerName;
        this.playerID = playerID;
        this.playerList = playerList;
    }

    @Override
    public IPacketSender getSender()
    {
        return sender;
    }

    @Override
    public void handle()
    {
        var conn = (IConnectedClient) sender;
        var client = playerList.getUnauthorizedClient(conn);

        client.setName(playerName);
        client.setID(playerID);

        playerList.registerPlayer(client);

        logger.info("Authenticated player '{}' with id {}.", playerName, playerID);
    }

    @Override
    public long getPacketID()
    {
        return PACKET_ID;
    }

    @Override
    public void write(DataOutput out)
    {
        // Auth packets are not sent from server.
    }
}
